pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,907 @@
+	// Check volume for provisioning requirements using its bound claim.
+	claim := obj.(*api.PersistentVolumeClaim)
+
+	// security check -- claim might not yet have been matched to the volume.
+	if pv.Name != claim.Spec.VolumeName && pv.Annotations[provisionedForKey] != ClaimToProvisionableKey(claim) {
+		glog.V(5).Infof("Security mismatch.  Expecting %s but found %s", pv.Name, claim.Spec.VolumeName)
+		return pv, api.PersistentVolumeStatus{
+			Phase:   api.VolumeFailed,
+			Message: "Mismatched claim/volume names",
+		}, fmt.Errorf("Mismatched claim/volume names - expected %s but found %s", claim.Spec.VolumeName, pv.Name)
+	}
+
+	// no provisioning required, volume is ready and Bound
+	if !keyExists(pvProvisioningRequired, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] is bound and no provisioning is required", claim.Spec.VolumeName)
+		return pv, api.PersistentVolumeStatus{Phase: api.VolumeBound}, nil
+	}
+
+	// provisioning is completed, volume is ready and return Bound
+	if isAnnotationMatch(pvProvisioningRequired, pvProvisioningCompleted, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] is bound and provisioning is complete", pv.Name)
+		provisionedFor, _ := pv.Annotations[provisionedForKey]
+		if provisionedFor != ClaimToProvisionableKey(claim) {
+			return pv, api.PersistentVolumeStatus{
+				Phase:   api.VolumeFailed,
+				Message: "Mismatched claim/volume annotations",
+			}, fmt.Errorf("pre-bind mismatch - expected %s but found %s", ClaimToProvisionableKey(claim), provisionedFor)
+		}
+		return pv, api.PersistentVolumeStatus{Phase: api.VolumeBound}, nil
+	}
+
+	// provisioning is not complete, return Pending
+	if !isAnnotationMatch(pvProvisioningRequired, pvProvisioningCompleted, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] - provisioning in progress", pv.Name)
+		// one lock per volume to allow parallel processing but limit activity per volume
+		if _, exists := controller.locks[pv.Name]; !exists {
+			controller.locks[pv.Name] = &sync.RWMutex{}
+			go provisionVolume(pv, controller)
+		}
+
+		return pv, api.PersistentVolumeStatus{Phase: api.VolumePending, Message: "Awaiting provisioning"}, nil
+	}
+
+	return pv, api.PersistentVolumeStatus{Phase: api.VolumeFailed, Message: "Unknown state"}, nil
+}
+
+// provisionVolume provisions a volume that has been created in the cluster but not yet fulfilled by
+// the storage provider.  This func (and Recycle/Delete) locks on the PV.Name to limit 1 operation per volume at a time.
+func provisionVolume(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+	controller.locks[pv.Name].Lock()
+	defer func(pv *api.PersistentVolume, controller *PersistentVolumeController) {
+		controller.locks[pv.Name].Unlock()
+		delete(controller.locks, pv.Name)
+	}(pv, controller)
+
+	if isAnnotationMatch(pvProvisioningRequired, pvProvisioningCompleted, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] is already provisioned", pv.Name)
+	}
```

Comment:
How do errors from this function propagate to the volume and the user behind it? There  is just glog.V(5) everywhere.